File: ios/App/App/OBubba/Plugins/ReviewPlugin.swift
import Foundation
import Capacitor
import StoreKit

/// Capacitor plugin for native in-app review prompts.
/// Uses SKStoreReviewController on iOS — Apple controls frequency
/// (max 3 times per 365-day period) and won't show if user already reviewed.
@objc(OBReview)
public class ReviewPlugin: CAPPlugin, CAPBridgedPlugin {
    public let identifier = "OBReview"
    public let jsName = "OBReview"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "requestReview", returnType: CAPPluginReturnPromise),
    ]

    @objc func requestReview(_ call: CAPPluginCall) {
        DispatchQueue.main.async {
            if #available(iOS 16.0, *) {
                if let scene = UIApplication.shared.connectedScenes
                    .first(where: { $0.activationState == .foregroundActive }) as? UIWindowScene {
                    SKStoreReviewController.requestReview(in: scene)
                }
            } else if #available(iOS 14.0, *) {
                if let scene = UIApplication.shared.connectedScenes.first as? UIWindowScene {
                    SKStoreReviewController.requestReview(in: scene)
                }
            }
            call.resolve(["requested": true])
        }
    }
}

File: android/app/src/main/java/com/obubba/app/plugins/ReviewPlugin.java
package com.obubba.app.plugins;

import android.app.Activity;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.JSObject;
import com.getcapacitor.annotation.CapacitorPlugin;
import com.google.android.play.core.review.ReviewInfo;
import com.google.android.play.core.review.ReviewManager;
import com.google.android.play.core.review.ReviewManagerFactory;

/**
 * Capacitor plugin for Google Play In-App Review.
 * Google controls frequency and won't show if user already reviewed.
 */
@CapacitorPlugin(name = "OBReview")
public class ReviewPlugin extends Plugin {

    @PluginMethod
    public void requestReview(PluginCall call) {
        Activity activity = getActivity();
        if (activity == null) {
            JSObject ret = new JSObject();
            ret.put("requested", false);
            call.resolve(ret);
            return;
        }

        ReviewManager manager = ReviewManagerFactory.create(activity);
        manager.requestReviewFlow().addOnCompleteListener(task -> {
            if (task.isSuccessful()) {
                ReviewInfo reviewInfo = task.getResult();
                manager.launchReviewFlow(activity, reviewInfo).addOnCompleteListener(flowTask -> {
                    JSObject ret = new JSObject();
                    ret.put("requested", true);
                    call.resolve(ret);
                });
            } else {
                JSObject ret = new JSObject();
                ret.put("requested", false);
                call.resolve(ret);
            }
        });
    }
}
